pinot-core/src/main/java/org/apache/pinot/core/query/executor/MdcExecutor.java

```diff
@@ -0,0 +1,159 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.executor;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+
+/**
+ * A delegator executor service that sets MDC context for the query.
+ *
+ * By using this executor, all tasks submitted to the executor will have the MDC context set to the query context.
+ * This is easier and safer to apply than setting the MDC context manually for each task.
+ * You can read more about MDC in <a href="https://www.baeldung.com/mdc-in-log4j-2-logback">this baeldung post</a>
+ *
+ * TODO: Convert this class and its usages into an Executor instead of an ExecutorService
+ *
+ */
+public abstract class MdcExecutor implements ExecutorService {
+  private final ExecutorService _executorService;
+
+  public MdcExecutor(ExecutorService executorService) {
+    _executorService = executorService;
+  }
+
+  protected abstract boolean alreadyRegistered();
```

Comment:
At least with the current implementation, this fn. does not have to be abstract ? Both the implementations are the same.